include/boost/json/impl/monotonic_resource.ipp

```diff
@@ -79,35 +97,69 @@ monotonic_resource::
 }
 
 monotonic_resource::
-monotonic_resource() noexcept
-    : block_size_(64 * 1024)
+monotonic_resource(
+    std::size_t max_alloc,
+    std::size_t initial_size) noexcept
+      : max_alloc_(max_alloc),
+        block_size_(initial_size)
 {
 }
 
 monotonic_resource::
 monotonic_resource(
-    std::size_t block_size) noexcept
-    : block_size_(block_size)
+  void* buffer,
+  std::size_t buffer_size,
+  std::size_t max_alloc) noexcept
+    : initial_(buffer, buffer_size),
+      max_alloc_(max_alloc),
+      block_size_(1024)
 {
 }
 
+void
+monotonic_resource::
+clear() noexcept
+{
+    // reset the initial block (if any)
+    if (initial_.base)
+      initial_.top = initial_.base_ptr();
+    // reset all the blocks managed
+    // by this object (if any)
+    for (auto b = head_; b;)
+    {
+      auto next = b->next;
+      b->top = b->base_ptr();
+      b = next;
+    }
+    allocated_ = 0;
+}
+
 void*
 monotonic_resource::
 do_allocate(
     std::size_t n,
     std::size_t align)
 {
+    // use the initial block if possible
+    if(initial_.base)
+        if (auto p = alloc_in_block(initial_, n, align))
+            return p;
+    // cap the number of dynamically allocated bytes
+    if(allocated_ + n > max_alloc_)
+        throw std::bad_alloc();
+    // do this after the check for exception guarantee
+    allocated_ += n;
     if(head_)
-    {
-        auto p = head_->alloc(n, align);
-        if(p)
+        if(auto p = alloc_in_block(*head_, n, align))
             return p;
-    }
-    if(n > block_size_ - 2 * align)
-        alloc_block(n + 2 * align);
-    else
-        alloc_block(block_size_);
-    auto p = head_->alloc(n, align);
+    // if the requested size is greater than the block size,
+    // use the requested size instead
+    const auto new_block_size = n > block_size_ - 2 * align ?
+        n + 2 * align : block_size_;
```

Comment:
Best not the use the ternary operator here, because of code coverage. It isn't quite as tidy but this is better:
```
std::size_t new_size;
if( n > block_size_ - 2 * align )
  new_size = n + 2 * align;
else
  new_size = block_size_;
```

Also are we checking that `n + 2 * align` does not overflow?

Reply:
Would an IILE work?

Reply:
ugh too try-hard